lib-find-meal/src/main/java/com/kinandcarta/lib/find/meal/viewmodel/FindMealViewModel.kt

```diff
@@ -1,7 +1,86 @@
 package com.kinandcarta.lib.find.meal.viewmodel
 
+import android.location.Location
+import androidx.annotation.RequiresPermission
+import androidx.hilt.lifecycle.ViewModelInject
+import androidx.lifecycle.LiveData
+import androidx.lifecycle.MutableLiveData
 import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import com.google.android.gms.location.FusedLocationProviderClient
+import com.google.android.gms.location.LocationCallback
+import com.google.android.gms.location.LocationRequest
+import com.google.android.gms.location.LocationResult
+import com.kcc.kmmhackathon.shared.MealsSDK
+import com.kcc.kmmhackathon.shared.entity.Meal
+import com.kcc.kmmhackathon.shared.utility.DistanceUnit
+import kotlinx.coroutines.launch
 
-class FindMealViewModel : ViewModel() {
-    // TODO: Implement the ViewModel
+typealias Meals = List<Meal>
+
+class FindMealViewModel @ViewModelInject constructor(
+    private val fusedLocationClient: FusedLocationProviderClient
+) : ViewModel() {
+
+    sealed class State {
+        object LoadingMeals : State()
+        data class LoadedMeals(val meals: Meals, val distanceUnit: DistanceUnit) : State()
+        data class Failed(val failure: Failure) : State()
+    }
+
+    sealed class Failure(cause: Throwable) : Throwable(cause) {
+        class LoadingMealsFailed(cause: Throwable) : Failure(cause)
+    }
+
+    val state: LiveData<State> get() = _state
+
+    private val _state: MutableLiveData<State> = MutableLiveData(State.LoadingMeals)
+    private var lastLocation: Location? = null
+    private val sdk = MealsSDK()
+    private val locationCallback = object : LocationCallback() {
+        override fun onLocationResult(locationResult: LocationResult?) {
+            lastLocation = locationResult?.lastLocation
+            updateMeals(true) // TODO: Check param value
+        }
+    }
+
+    override fun onCleared() {
+        fusedLocationClient.removeLocationUpdates(locationCallback)
+    }
+
+    @RequiresPermission("android.permission.ACCESS_FINE_LOCATION")
+    fun startUpdatingLocation() {
+        fusedLocationClient.lastLocation.addOnSuccessListener {
+            lastLocation = it
+            updateMeals(true)
+        }
+        fusedLocationClient.requestLocationUpdates(
+            createLocationRequest(),
+            locationCallback,
+            null
+        )
+    }
+
+    fun updateMeals(needReload: Boolean) {
+        val location = lastLocation ?: return
+        val distanceUnit = DistanceUnit.miles
+        _state.value = State.LoadingMeals
+        viewModelScope.launch {
```

Comment:
Android VMs come with a scope that automatically gets cancelled on clear I believe